### ISSUE TITLE
Respeaker 4 hat

### ISSUE THREAD

**Issue by pointout** [createcandle/voco#6]
I have a respeaker 4 hat and a usb-speaker for a raspberry pi 3+ with snips. That works fine. I would like to use this hardware for mozilla-iot with voco. Is there a way to do this?

**Comment by pointout** [comment]
I have made a
```
git clone https://github.com/respeaker/seeed-voicecard
cd seeed-voicecard
sudo ./install.sh
```
and on sam a 
`sam setup audio`
I could test the mircrophone and speaker with sam, but voco only shows under microphone: "Attached deivise (1/0)"  in the configuration. If I select this I get no confirmation tone for the "hey snips" command. 
If I select "Built-in  icrophone (0,0)", I get an Error: undefined 0.
The usb speaker is saying something containing "snips..." on startup, but I cannot understand the rest. (setting "Attached device (1,0)").

**Comment by pointout** [comment]
I think It would need changes in manifest.json and /pkg/voco_device.py. ("Respeacker array (2,0)"). 
I am not familiar in how to pack and test the addon. 
(I got it working with a usb soundcard and a microphone, but would really like to use the Respeaker array.)

**Comment by createcandle** [comment]
I've added an experimental option to support ReSpeaker hat. Would you be willing to test it?

**Comment by pointout** [comment]
Yes of course! How do I install it?

Edit: Just saw, that verison in add-on-list changed to 1.0.2 after removing 1.0.1

**Comment by pointout** [comment]
Installing does not finish.
```
2019-12-18 16:18:14.222 INFO   : Fetching add-on https://s3-us-west-2.amazonaws.com/mozilla-gateway-addons/voco-1.0.2.tgz as /tmp/3yX8Ex/voco.tar.gz
2019-12-18 16:20:14.226 INFO   : Fetching add-on https://s3-us-west-2.amazonaws.com/mozilla-gateway-addons/voco-1.0.2.tgz as /tmp/gV95JM/voco.tar.gz
2019-12-18 16:22:14.232 INFO   : Fetching add-on https://s3-us-west-2.amazonaws.com/mozilla-gateway-addons/voco-1.0.2.tgz as /tmp/LiSYM4/voco.tar.gz
```
Edit: ok it seems that it took longer, but finished now.

**Comment by pointout** [comment]
Mostly good news:
The microphone worked fine just setting microphone to "ReSpeaker" on the configuration page. The usb speaker I had to set with sam. Before setting it with sam the wake word confirmation sound came out of the usb speaker, but the speach answers came from the build in headphone jack.

### Hardware:

- Raspberry Pi 3 B+
- [ReSpeacker 4 Mic Array](https://respeaker.io/4_mic_array/)
- [Adafruit mini USB Speaker](https://www.adafruit.com/product/3369)
- Conbee II Zigbee USB-stick
- IKEA TRÅDFRI Wireless Control Outlet

### What I did:

- Pluged in the Speaker and microphone
- installed mozilla IoT gateway
- did't connect it to mozilla-iot.org for the moment
- enabeld ssh and installed respeaker driver:
```
$ git clone https://github.com/respeaker/seeed-voicecard.git
$ cd seeed-voicecard
$ sudo ./install.sh
$ sudo reboot
```
- installed voco add on 
-- inserted Authorization token (developer-create local authorization) and selected:
-- microphone: ReSpeaker (2,0)
-- speaker: Attached device (1,0)
at this point commands are working, but speach answers only came from the headphone jack, while wake word confirmation sound came from usb speaker.
- logged in with snips "sam" `sam connect gateway.local`
- `sam setup audio` and select `card 1: Device [USB2.0 Device], device 0: USB Audio [USB Audio]`

Microphone and usb-speaker are working. I can turn on and off the zigbee IKEA TRÅDFRI Wireless Control Outlet and set timers, alarms and countdowns and ask for the time.

**Comment by pointout** [comment]
It seems that adjusting the volume is not working for the USB speaker. It is hard to understand the voice and I think it would be better to understand, if the volume setting would be lower.

**Comment by pointout** [comment]
I changed to JBL Go Speaker. The speaker is powered over an usb port from the raspi, so only the raspi power supply is needed. Audio comes from the 3,5mm audio jack. If snips is set to 100%, the speaker is not powering off. The answers are much better understandable than with the Adafruit Speaker.

**Comment by createcandle** [comment]
Thank you for your help!

> -- speaker: Attached device (1,0)

Out of curiosity, do you alse see `reSpeaker` as an output option? Although that will probably just select the reSpeaker's 3,5mm output (if it has that).

Looking at the code, I believe changing to USB output simply isn't implemented yet:
https://github.com/createcandle/voco/blob/master/pkg/voco_adapter.py#L1373

But that can change :-)

It's most likely just be a matter of setting the USB soundcard as the default audio output, as `nanotts` will speak to whatever is set as the default.

Perhaps you can try? Log in with SSH again and run `alsamixer`. Then, try pressing `F6` and see if that allows you to set the default audio output? (I don't know if that will work).

This also looks relevant, and would be what could be implemented into the add-on. If this works for you that would be good news, as I could get the add-on to do this automatically in the future. Let me know and I will add it.
https://unix.stackexchange.com/questions/88118/alsa-how-to-temporarily-change-the-output-device

Another more permanent option is to try this: https://www.alsa-project.org/main/index.php/Setting_the_default_device

But Gateway Addons are not allowed to make such changes to settings on the system. But it might be useful for you as a quick fix anyway, as it will keep the settings after a reboot.

> If snips is set to 100%, the speaker is not powering off.

I don't quite follow. The volume level influences whether the speaker goes into some kind of sleep mode? It does work if it's set to 99%?

**Comment by pointout** [comment]
I did a clean new Mozilla IoT Gateway installation with the voco add on. Switching output from "build-in headphone jack (0,0)" to "Attached device (1,0)" (=USB Speaker or Soundcard) and back again is working perfect. 

> Out of curiosity, do you alse see reSpeaker as an output option? Although that will probably just select the reSpeaker's 3,5mm output (if it has that).

Yes I see reSpeaker as an output option. My Respeaker 4 Mic Array has no headphone jack.

I installed the repseaker driver:
```
$ git clone https://github.com/respeaker/seeed-voicecard.git
$ cd seeed-voicecard
$ sudo ./install.sh
$ sudo reboot
```
and selected the respeaker option in voco configuration, but got an error:
`undefined 512 `

Then I connect to gateway with sam an make:
`sam setup audio`
sam finds the respeaker microphone and than the microphone works.

The links you mentioned are for the USB speaker, am I right? 



>> If snips is set to 100%, the speaker is not powering off.

>I don't quite follow. The volume level influences whether the speaker goes into some kind of sleep mode? It does work if it's set to 99%?

The first time I used this speaker with a lower output level it turned off automatically. Now I had the output level at 90% for some minutes and the speaker is still on. I maybe will test this more precisely.

**Comment by createcandle** [comment]
It's surprising that `sam` works, since technically Snips is not installed on the Gateway. Snips has been modified to run standalone from a directory, without needing it be be installed.

**Comment by createcandle** [comment]
I just got a chance to test a ReSpeaker Pi Hat, the cheap 2 microphone version. After installing the driver and selected ReSpeaker in the Voco settings it worked immediately.

The only issue is that the noises to indicate it heard a command cannot be heard. They are probably being played over the normal 3,5 inch output jack. That should be easy to fix.

**Comment by pointout** [comment]
Does the small microhead have a build in speaker? I allways heard the confirmation sound over the speaker at the 3,5 output jack.

I just installed a fresh 0.12 mozilla gateway.
Than I installed the (latest) seeed-voicecard driver. After that I have only the option for 3,5 outpout jack. No usb. The Microphone is working, but is also the only option.
If I go to the Settings page of voco, the microphone is set to "Attached device (1,0)". Changing to respeaker and saving and going back to settings shows again "Attached device". It is working, but the output seems to be not as loud as before.
There is also a ticking on the speaker while it usb plug is plugged into the raspi for loading. If I plug of usb the ticking is gone.
